packages/runtime-core/src/component.ts

```diff
@@ -219,11 +221,25 @@ export const setCurrentInstance = (
   currentInstance = instance
 }
 
+export function validateComponentName(name: string, config: AppConfig) {
+  const appIsNativeTag = config.isNativeTag || NO
+  if (isBuiltInTag(name) || appIsNativeTag(name)) {
+    warn(
+      'Do not use built-in or reserved HTML elements as component id: ' + name
+    )
+  }
+}
+
 export function setupStatefulComponent(
   instance: ComponentInternalInstance,
   parentSuspense: SuspenseBoundary | null
 ) {
   const Component = instance.type as ComponentOptions
+
+  if (__DEV__ && Component.name) {
+    validateComponentName(Component.name, instance.appContext.config)
```

Comment:
Should also validate all names in `Component.components`

Reply:
Done.